docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,636 @@
+        return {layer: layer._get_slice(request) for layer, request in requests.items()}
+
+    def _on_slice_done(self, task: Future[ViewerSliceResponse]) -> None:
+        if task.cancelled():
+            return
+        self.events.slice_ready(task.result())
+```
+
+For this class to be useful, there should be at least one connection to the `slice_ready` event.
+In napari, we expect the `QtViewer` to marshall the slice response that this signal carries
+to the vispy layers so that the canvas can be updated.
+
+### Hooking up the viewer
+
+Using Python's standard library threads in the `ViewerModel`
+mean that we have a portable way to perform asynchronous slicing
+in napari without an explicit dependency on Qt.
+
+```python
+class ViewerModel:
+    ...
+
+    dims: Dims
+    _slicer: LayerSlicer = LayerSlicer()
+
+    def __init__(self, ...):
+        ...
+        self.dims.events.current_step.connect(self._slice_layers_async)
+
+    ...
+
+    def _slice_layers_async(self) -> None:
+        self._slicer.slice_layers_async(self.layers, self.dims)
+```
+
+The main response to the slice being ready occurs on the `QtViewer`
+because `QtViewer.layer_to_visual` provides a way to map from
+a layer to its corresponding vispy layer.
+
+```python
+class QtViewer:
+    ...
+
+    viewer: ViewerModel
+    layer_to_visual: Dict[Layer, VispyBaseLayer]
+
+    def __init__(self, ...):
+        ...
+        self.viewer._slicer.events.slice_ready.connect(self._on_slice_ready)
+
+    @ensure_main_thread
+    def _on_slice_ready(self, responses: ViewerSliceResponse):
+        for layer, response in responses.items():
+            if visual := self.layer_to_visual[layer]:
+                visual._set_slice(response)
+```
+
+The other reason is because `QtViewer` is a `QObject` that lives in the main
+thread, which makes it easier to ensure that the slice response is handled
+on the main thread.
+That's useful because Qt widgets and vispy nodes can only be safely updated
+on the main thread [^vispy-faq-threads], both of which occur when consuming
+slice output.
+
+
+## Implementation
+
+A tracking issue [^tracking-issue] serves as a way to communicate high level
+progress and discussion.
+
+A prototype [^prototype-pr] has been created that implements the approach described
+above to show its basic technical feasibility in napari's existing code base.
+It only implements it for some layer types, some slicing operations, and may
+not work when layer data is mutated.
+There are some other rough edges and plenty of ways to break it, but basic
+functionality should work.
+
+
+## Backward compatibility
+
+### Breaks synchronous slicing behavior
+
+The main goal of this project is to perform slicing asynchronously, so it's
+natural that we might break anyone that was depending on slicing being synchronous.
+At a minimum, we must provide a public way to achieve the same fundamental goals,
+such as connecting to the `slice_ready` event.
```

Comment:
Writing a project that uses napari to create animation having only `slice_ready` will be much harder than for the current state as it will require a switch from some form of only loop to object keeping state to pass around waiting for the event. It may be nice to have fully sync mode. 

@alisterburt I think that it may touch your work.

Reply:
napari animation is a great call out, we'd have to make that respond to some event after rendering is done to make sure we captured the screenshot at the right time. we might need an additional event then or a force sync mode

Reply:
Great point @Czaki !

Reply:
I think you mentioned this verbally, but thanks for writing it down. I think it would be a useful exercise to try to see how hard it is to make napari-animation work with the current prototype in: https://github.com/andy-sweet/napari/pull/21

I'm planning to do that, but won't complain if someone else gets to it before me!